# Counterfactuals Training Data Extraction Experiment

In [8]:
import pandas as pd
from sklearnex import patch_sklearn
patch_sklearn()
import sklearn.ensemble as es
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
import random
import logging
import warnings
import dice_ml

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


In [9]:
threads = 15

logging.basicConfig()

logger = logging.getLogger('xai-privacy')

In [10]:
from experiment_setup import run_all_experiments
from experiment_setup import get_heart_disease_dataset
from experiment_setup import get_census_dataset

In [11]:
DATASET_HALF = False

data_heart_dict, data_heart_num_dict, data_heart_cat_dict = get_heart_disease_dataset(halve_dataset=DATASET_HALF)
data_census_dict, data_census_num_dict, data_census_cat_dict = get_census_dataset(halve_dataset=DATASET_HALF)

data_heart = data_heart_dict['dataset']
data_heart_num = data_heart_num_dict['dataset']
data_heart_cat = data_heart_cat_dict['dataset']
data_census = data_census_dict['dataset']
data_census_num = data_census_num_dict['dataset']
data_census_cat = data_census_cat_dict['dataset']
outcome_name_heart = data_heart_dict['outcome']
numeric_features_heart = data_heart_dict['num']

Feature Age: removed 0 rows for missing values.
Feature RestingBP: removed 59 rows for missing values.
Feature Cholesterol: removed 27 rows for missing values.
Feature FastingBS: add unknown category 2.0
Feature RestingECG: add unknown category 3.0
Feature MaxHR: removed 0 rows for missing values.
Feature Oldpeak: removed 7 rows for missing values.
Feature ST_Slope: add unknown category 4.0
Feature CA: add unknown category 4.0
Feature Thal: add unknown category 8.0
Dropped 271 of 1097
Dropped 273 of 1097
Dropped 277 of 1097
Dropped: 2399 of 32561
census: Dropped 3848 of 30162
num: Dropped 19859 of 30162
cat: Dropped 12136 of 30162


This notebook will test whether training data extraction is possible with counterfactuals (CF) that are drawn from the training data. Training data extraction means an attacker can find out the feature values of samples from the training data without prior knowledge of them. The attacker only has access to the model's prediction function and the explanation.

This attack should be trivial because any counterfactual that is shown as an explanation was picked directly from the training data.

The idea for counterfacutal training data extraction is as follows: The attacker makes repeated queries to the model with random input values. In order to do this, the attacker knows the maximum and minimum value of each feature in the training data (or the categorical values of each feature). The returned counterfactuals are the extracted training data.

First, we implement the `train_explainer` and `training_data_extraction_model_access` functions:

In [12]:
# Attack code must be imported so that multiprocessing pool works. Check out ice_attack.py for the implementation of the attack.
from cf_attack import CounterfactualTDE

# Executing Training Data Extraction

We now generate five counterfactuals for the first sample from the training data to demonstrate counterfactual explanations in general.

In [13]:
features = data_heart.drop(outcome_name_heart, axis=1)
labels = data_heart[outcome_name_heart]

# Train a random forest on training data.
model = es.RandomForestClassifier(random_state=0)
model = model.fit(features, labels)

# Train explainer
d = dice_ml.Data(dataframe=data_heart, continuous_features=numeric_features_heart, outcome_name=outcome_name_heart)

m = dice_ml.Model(model=model, backend="sklearn", model_type='classifier')
# Generating counterfactuals from training data (kd-tree)
exp = dice_ml.Dice(d, m, method="kdtree")

In [14]:
e1 = exp.generate_counterfactuals(features[0:1], total_CFs=5, desired_class="opposite")
e1.visualize_as_dataframe(display_sparse_df=False)

100%|██████████| 1/1 [00:00<00:00,  1.04it/s]

Query instance (original outcome : 0)


Age  Sex ChestPainType  RestingBP  Cholesterol FastingBS RestingECG  \
0  47.0  1.0           1.0      110.0        249.0       0.0        0.0   

   MaxHR ExerciseAngina  Oldpeak ST_Slope   CA Thal  HeartDisease  
0  150.0            0.0      0.0      4.0  4.0  8.0           0.0


Diverse Counterfactual set without sparsity correction (new outcome:  1


Age  Sex ChestPainType  RestingBP  Cholesterol FastingBS RestingECG  \
388   35.0  1.0           2.0      110.0        257.0       0.0        0.0   
434   47.0  1.0           3.0      108.0        243.0       0.0        0.0   
705   54.0  1.0           3.0      120.0        237.0       0.0        0.0   
320   46.0  1.0           4.0      120.0        249.0       0.0        2.0   
1111  33.0  0.0           4.0      100.0        246.0       0.0        0.0   

      MaxHR ExerciseAngina  Oldpeak ST_Slope   CA Thal  
388   140.0            0.0      0.0      4.0  4.0  8.0  
434   152.0            0.0      0.0      1.0  0.0  3.0  
705   150.0            1.0      1.5      4.0  4.0  7.0  
320   144.0            0.0      0.8      1.0  0.0  7.0  
1111  150.0            1.0      1.0      2.0  4.0  8.0

We can see that the counterfactuals are similar to the query sample and that they have a flipped prediction. These are the two general properties of counterfactual explanations.

We will now do a small proof of concept of the experiment with logging enabled to demonstrate how it works.

In [15]:
logger.setLevel(logging.DEBUG)
logging.root.setLevel(logging.ERROR)

EXP = CounterfactualTDE(data_heart, numeric_features_heart, outcome_name_heart, random_state=0)
EXP.training_data_extraction_experiment(num_queries=12, model=es.RandomForestClassifier(random_state=0), model_access=False)

logger.setLevel(logging.INFO)

DEBUG:xai-privacy:Numeric Features: ['Age', 'RestingBP', 'Cholesterol', 'MaxHR', 'Oldpeak']
DEBUG:xai-privacy:Categorical Features: ['CA', 'ChestPainType', 'ExerciseAngina', 'FastingBS', 'RestingECG', 'ST_Slope', 'Sex', 'Thal']
100%|██████████| 12/12 [00:09<00:00,  1.29it/s]
DEBUG:xai-privacy:Sample 0: Counterfactuals 
 [[38.0 '1.0' '4.0' 92.0 117.0 '0.0' '0.0' 134.0 '1.0' 2.5 '2.0' '4.0'
  '8.0']
 [57.0 '1.0' '4.0' 130.0 131.0 '0.0' '0.0' 115.0 '1.0' 1.2 '2.0' '1.0'
  '7.0']
 [51.0 '0.0' '4.0' 120.0 0.0 '2.0' '0.0' 127.0 '1.0' 1.5 '1.0' '4.0'
  '8.0']
 [36.0 '1.0' '4.0' 110.0 0.0 '2.0' '0.0' 125.0 '1.0' 1.0 '2.0' '4.0'
  '6.0']]
DEBUG:xai-privacy:Sample 1: Counterfactuals 
 [[48.0 '1.0' '4.0' 160.0 268.0 '0.0' '0.0' 103.0 '1.0' 1.0 '2.0' '4.0'
  '8.0']
 [57.0 '1.0' '4.0' 150.0 276.0 '0.0' '2.0' 112.0 '1.0' 0.6 '2.0' '1.0'
  '6.0']
 [57.0 '1.0' '4.0' 165.0 289.0 '1.0' '2.0' 124.0 '0.0' 1.0 '2.0' '3.0'
  '7.0']
 [56.0 '0.0' '4.0' 200.0 288.0 '1.0' '2.0' 133.0 '1.0' 4.0 '3.0' '2.0'
  '7.

Total time: 9.74s (training model: 0.05s, training explainer: 0.05s, experiment: 9.64s)
Number of extracted samples: 41
Number of accurate extracted samples: 41
Precision: 1.0, recall: 3.4166666666666665


The proof of concept should show that each extracted sample is an actual training sample (precision of 100%). Recall is above 100% because this method can extract multiple samples per query (multiple counterfactuals are returned). Recall will reach a reasonable value if the experiment is executed for the full training data. In this case, the attack cannot return more samples than the number of queries because the attack is limited by the number of training samples.

Now we begin executing the actual experiment. We begin by defining the table that will hold the results for all our different experiment variations. Then we execute all variations of the experiment for this dataset. We vary the model between a decision tree, a random forest and a neural network. Each model uses the default configuration of scikit-learn.

In [16]:
dataset_dicts = [data_heart_dict, data_heart_num_dict, data_heart_cat_dict, data_census_dict, data_census_num_dict, data_census_cat_dict]

dt_dict = {'name': 'decision tree', 'model': DecisionTreeClassifier}
rf_dict = {'name': 'random forest', 'model': es.RandomForestClassifier}
nn_dict = {'name': 'neural network', 'model': MLPClassifier}

model_dicts = [dt_dict, rf_dict, nn_dict]

# We set the number of extractions to the length of the dataset
num_queries_dict = { 'heart': len(data_heart), 'heart numeric': len(data_heart_num), 'heart categorical': len(data_heart_cat), 'census': len(data_census), 'census numeric': len(data_census_num), 'census categorical': len(data_census_cat)}

In [17]:
# remove pandas warnings
warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)

In [ ]:
# This will run the experiment for each dataset and model combination

results = run_all_experiments(CounterfactualTDE, dataset_dicts, model_dicts, random_state=0, num_queries=num_queries_dict, model_access=False, threads=threads, is_mem_inf=False, convert_cat_to_str=True, repeat=10)

dataset: heart, model: decision tree (repetition 0)
Total time: 72.08s (training model: 0.02s, training explainer: 0.04s, experiment: 72.03s)
Number of extracted samples: 529
Number of accurate extracted samples: 529
Precision: 1.0, recall: 0.6404358353510896
dataset: heart, model: decision tree (repetition 1)
Total time: 70.20s (training model: 0.01s, training explainer: 0.03s, experiment: 70.16s)
Number of extracted samples: 536
Number of accurate extracted samples: 536
Precision: 1.0, recall: 0.648910411622276
dataset: heart, model: decision tree (repetition 2)
Total time: 65.65s (training model: 0.02s, training explainer: 0.03s, experiment: 65.60s)
Number of extracted samples: 506
Number of accurate extracted samples: 506
Precision: 1.0, recall: 0.612590799031477
dataset: heart, model: decision tree (repetition 3)
Total time: 64.55s (training model: 0.02s, training explainer: 0.02s, experiment: 64.51s)
Number of extracted samples: 558
Number of accurate extracted samples: 558
Preci

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 69.01s (training model: 1.95s, training explainer: 0.01s, experiment: 67.05s)
Number of extracted samples: 466
Number of accurate extracted samples: 466
Precision: 1.0, recall: 0.5641646489104116
dataset: heart, model: neural network (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 67.93s (training model: 1.90s, training explainer: 0.01s, experiment: 66.02s)
Number of extracted samples: 419
Number of accurate extracted samples: 419
Precision: 1.0, recall: 0.5072639225181598
dataset: heart, model: neural network (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 66.52s (training model: 1.92s, training explainer: 0.01s, experiment: 64.60s)
Number of extracted samples: 435
Number of accurate extracted samples: 435
Precision: 1.0, recall: 0.5266343825665859
dataset: heart, model: neural network (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 69.43s (training model: 1.88s, training explainer: 0.00s, experiment: 67.54s)
Number of extracted samples: 435
Number of accurate extracted samples: 435
Precision: 1.0, recall: 0.5266343825665859
dataset: heart, model: neural network (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 67.23s (training model: 1.94s, training explainer: 0.01s, experiment: 65.28s)
Number of extracted samples: 464
Number of accurate extracted samples: 464
Precision: 1.0, recall: 0.5617433414043583
dataset: heart, model: neural network (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 65.40s (training model: 1.96s, training explainer: 0.02s, experiment: 63.42s)
Number of extracted samples: 438
Number of accurate extracted samples: 438
Precision: 1.0, recall: 0.5302663438256658
dataset: heart, model: neural network (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 67.45s (training model: 1.91s, training explainer: 0.01s, experiment: 65.52s)
Number of extracted samples: 468
Number of accurate extracted samples: 468
Precision: 1.0, recall: 0.5665859564164649
dataset: heart, model: neural network (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 69.05s (training model: 1.93s, training explainer: 0.00s, experiment: 67.12s)
Number of extracted samples: 461
Number of accurate extracted samples: 461
Precision: 1.0, recall: 0.5581113801452785
dataset: heart, model: neural network (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 68.83s (training model: 1.92s, training explainer: 0.01s, experiment: 66.91s)
Number of extracted samples: 486
Number of accurate extracted samples: 486
Precision: 1.0, recall: 0.5883777239709443
dataset: heart, model: neural network (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 67.25s (training model: 1.86s, training explainer: 0.01s, experiment: 65.37s)
Number of extracted samples: 454
Number of accurate extracted samples: 454
Precision: 1.0, recall: 0.549636803874092
dataset: heart numeric, model: decision tree (repetition 0)
Total time: 32.25s (training model: 0.02s, training explainer: 0.00s, experiment: 32.24s)
Number of extracted samples: 547
Number of accurate extracted samples: 547
Precision: 1.0, recall: 0.6638349514563107
dataset: heart numeric, model: decision tree (repetition 1)
Total time: 30.96s (training model: 0.00s, training explainer: 0.00s, experiment: 30.96s)
Number of extracted samples: 534
Number of accurate extracted samples: 534
Precision: 1.0, recall: 0.6480582524271845
dataset: heart numeric, model: decision tree (repetition 2)
Total time: 28.90s (training model: 0.00s, training explainer: 0.00s, experiment: 28.90s)
Number of extracted samples: 550
Number of accurate extracted samples: 550
Precision: 1.0, recall: 0.667475

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 31.53s (training model: 1.09s, training explainer: 0.00s, experiment: 30.45s)
Number of extracted samples: 496
Number of accurate extracted samples: 496
Precision: 1.0, recall: 0.6019417475728155
dataset: heart numeric, model: neural network (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 32.16s (training model: 1.17s, training explainer: 0.00s, experiment: 31.00s)
Number of extracted samples: 493
Number of accurate extracted samples: 493
Precision: 1.0, recall: 0.5983009708737864
dataset: heart numeric, model: neural network (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 32.26s (training model: 1.09s, training explainer: 0.00s, experiment: 31.17s)
Number of extracted samples: 501
Number of accurate extracted samples: 501
Precision: 1.0, recall: 0.6080097087378641
dataset: heart numeric, model: neural network (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 30.91s (training model: 1.07s, training explainer: 0.02s, experiment: 29.82s)
Number of extracted samples: 473
Number of accurate extracted samples: 473
Precision: 1.0, recall: 0.5740291262135923
dataset: heart numeric, model: neural network (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 32.10s (training model: 1.11s, training explainer: 0.00s, experiment: 30.98s)
Number of extracted samples: 518
Number of accurate extracted samples: 518
Precision: 1.0, recall: 0.6286407766990292
dataset: heart numeric, model: neural network (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 30.16s (training model: 1.08s, training explainer: 0.00s, experiment: 29.08s)
Number of extracted samples: 481
Number of accurate extracted samples: 481
Precision: 1.0, recall: 0.5837378640776699
dataset: heart numeric, model: neural network (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 32.05s (training model: 1.08s, training explainer: 0.00s, experiment: 30.97s)
Number of extracted samples: 489
Number of accurate extracted samples: 489
Precision: 1.0, recall: 0.5934466019417476
dataset: heart numeric, model: neural network (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 30.50s (training model: 1.08s, training explainer: 0.00s, experiment: 29.41s)
Number of extracted samples: 469
Number of accurate extracted samples: 469
Precision: 1.0, recall: 0.5691747572815534
dataset: heart numeric, model: neural network (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 29.00s (training model: 1.11s, training explainer: 0.00s, experiment: 27.89s)
Number of extracted samples: 493
Number of accurate extracted samples: 493
Precision: 1.0, recall: 0.5983009708737864
dataset: heart numeric, model: neural network (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 30.16s (training model: 1.10s, training explainer: 0.00s, experiment: 29.06s)
Number of extracted samples: 479
Number of accurate extracted samples: 479
Precision: 1.0, recall: 0.5813106796116505
dataset: heart categorical, model: decision tree (repetition 0)
Total time: 22.14s (training model: 0.03s, training explainer: 0.02s, experiment: 22.09s)
Number of extracted samples: 702
Number of accurate extracted samples: 702
Precision: 1.0, recall: 0.8560975609756097
dataset: heart categorical, model: decision tree (repetition 1)
Total time: 22.97s (training model: 0.02s, training explainer: 0.02s, experiment: 22.94s)
Number of extracted samples: 723
Number of accurate extracted samples: 723
Precision: 1.0, recall: 0.8817073170731707
dataset: heart categorical, model: decision tree (repetition 2)
Total time: 22.20s (training model: 0.02s, training explainer: 0.02s, experiment: 22.16s)
Number of extracted samples: 724
Number of accurate extracted samples: 724
Precision: 1.0, rec

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.07s (training model: 0.24s, training explainer: 0.03s, experiment: 33.80s)
Number of extracted samples: 644
Number of accurate extracted samples: 644
Precision: 1.0, recall: 0.7853658536585366
dataset: heart categorical, model: random forest (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.02s (training model: 0.19s, training explainer: 0.03s, experiment: 33.80s)
Number of extracted samples: 647
Number of accurate extracted samples: 647
Precision: 1.0, recall: 0.7890243902439025
dataset: heart categorical, model: random forest (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 33.96s (training model: 0.19s, training explainer: 0.04s, experiment: 33.74s)
Number of extracted samples: 650
Number of accurate extracted samples: 650
Precision: 1.0, recall: 0.7926829268292683
dataset: heart categorical, model: random forest (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.26s (training model: 0.20s, training explainer: 0.02s, experiment: 34.04s)
Number of extracted samples: 659
Number of accurate extracted samples: 659
Precision: 1.0, recall: 0.8036585365853659
dataset: heart categorical, model: random forest (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.06s (training model: 0.23s, training explainer: 0.02s, experiment: 33.80s)
Number of extracted samples: 659
Number of accurate extracted samples: 659
Precision: 1.0, recall: 0.8036585365853659
dataset: heart categorical, model: random forest (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 33.96s (training model: 0.20s, training explainer: 0.03s, experiment: 33.72s)
Number of extracted samples: 678
Number of accurate extracted samples: 678
Precision: 1.0, recall: 0.8268292682926829
dataset: heart categorical, model: random forest (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.00s (training model: 0.19s, training explainer: 0.03s, experiment: 33.78s)
Number of extracted samples: 653
Number of accurate extracted samples: 653
Precision: 1.0, recall: 0.7963414634146342
dataset: heart categorical, model: random forest (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.24s (training model: 0.21s, training explainer: 0.01s, experiment: 34.02s)
Number of extracted samples: 647
Number of accurate extracted samples: 647
Precision: 1.0, recall: 0.7890243902439025
dataset: heart categorical, model: random forest (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.07s (training model: 0.20s, training explainer: 0.03s, experiment: 33.84s)
Number of extracted samples: 669
Number of accurate extracted samples: 669
Precision: 1.0, recall: 0.8158536585365853
dataset: heart categorical, model: random forest (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 34.16s (training model: 0.22s, training explainer: 0.02s, experiment: 33.92s)
Number of extracted samples: 657
Number of accurate extracted samples: 657
Precision: 1.0, recall: 0.801219512195122
dataset: heart categorical, model: neural network (repetition 0)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.66s (training model: 1.56s, training explainer: 0.02s, experiment: 22.08s)
Number of extracted samples: 647
Number of accurate extracted samples: 647
Precision: 1.0, recall: 0.7890243902439025
dataset: heart categorical, model: neural network (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.64s (training model: 1.57s, training explainer: 0.02s, experiment: 22.05s)
Number of extracted samples: 663
Number of accurate extracted samples: 663
Precision: 1.0, recall: 0.8085365853658537
dataset: heart categorical, model: neural network (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.84s (training model: 1.60s, training explainer: 0.02s, experiment: 22.22s)
Number of extracted samples: 647
Number of accurate extracted samples: 647
Precision: 1.0, recall: 0.7890243902439025
dataset: heart categorical, model: neural network (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.72s (training model: 1.66s, training explainer: 0.02s, experiment: 22.05s)
Number of extracted samples: 655
Number of accurate extracted samples: 655
Precision: 1.0, recall: 0.7987804878048781
dataset: heart categorical, model: neural network (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.71s (training model: 1.60s, training explainer: 0.02s, experiment: 22.09s)
Number of extracted samples: 636
Number of accurate extracted samples: 636
Precision: 1.0, recall: 0.775609756097561
dataset: heart categorical, model: neural network (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 24.73s (training model: 1.60s, training explainer: 0.02s, experiment: 23.12s)
Number of extracted samples: 637
Number of accurate extracted samples: 637
Precision: 1.0, recall: 0.776829268292683
dataset: heart categorical, model: neural network (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.79s (training model: 1.66s, training explainer: 0.02s, experiment: 22.11s)
Number of extracted samples: 622
Number of accurate extracted samples: 622
Precision: 1.0, recall: 0.7585365853658537
dataset: heart categorical, model: neural network (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.79s (training model: 1.60s, training explainer: 0.02s, experiment: 22.17s)
Number of extracted samples: 628
Number of accurate extracted samples: 628
Precision: 1.0, recall: 0.7658536585365854
dataset: heart categorical, model: neural network (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.82s (training model: 1.83s, training explainer: 0.02s, experiment: 21.98s)
Number of extracted samples: 662
Number of accurate extracted samples: 662
Precision: 1.0, recall: 0.8073170731707318
dataset: heart categorical, model: neural network (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 23.72s (training model: 1.61s, training explainer: 0.02s, experiment: 22.09s)
Number of extracted samples: 648
Number of accurate extracted samples: 648
Precision: 1.0, recall: 0.7902439024390244
dataset: census, model: decision tree (repetition 0)
Total time: 1551.00s (training model: 0.93s, training explainer: 0.12s, experiment: 1549.95s)
Number of extracted samples: 1928
Number of accurate extracted samples: 1928
Precision: 1.0, recall: 0.07326898229079577
dataset: census, model: decision tree (repetition 1)
Total time: 1550.99s (training model: 0.88s, training explainer: 0.12s, experiment: 1549.99s)
Number of extracted samples: 2032
Number of accurate extracted samples: 2032
Precision: 1.0, recall: 0.07722125104507106
dataset: census, model: decision tree (repetition 2)
Total time: 1541.56s (training model: 0.95s, training explainer: 0.17s, experiment: 1540.43s)
Number of extracted samples: 1926
Number of accurate extracted samples: 1926
Precision: 1.0, recall: 0.073192

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3583.59s (training model: 16.73s, training explainer: 0.12s, experiment: 3566.73s)
Number of extracted samples: 1380
Number of accurate extracted samples: 1380
Precision: 1.0, recall: 0.05244356616249905
dataset: census, model: random forest (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3586.34s (training model: 16.85s, training explainer: 0.16s, experiment: 3569.34s)
Number of extracted samples: 1379
Number of accurate extracted samples: 1379
Precision: 1.0, recall: 0.05240556357832333
dataset: census, model: random forest (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3614.29s (training model: 16.75s, training explainer: 0.13s, experiment: 3597.42s)
Number of extracted samples: 1342
Number of accurate extracted samples: 1342
Precision: 1.0, recall: 0.05099946796382154
dataset: census, model: random forest (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3591.81s (training model: 16.71s, training explainer: 0.12s, experiment: 3574.97s)
Number of extracted samples: 1365
Number of accurate extracted samples: 1365
Precision: 1.0, recall: 0.051873527399863194
dataset: census, model: random forest (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3624.12s (training model: 17.03s, training explainer: 0.36s, experiment: 3606.73s)
Number of extracted samples: 1527
Number of accurate extracted samples: 1527
Precision: 1.0, recall: 0.05802994603633047
dataset: census, model: random forest (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3600.06s (training model: 16.55s, training explainer: 0.13s, experiment: 3583.38s)
Number of extracted samples: 1400
Number of accurate extracted samples: 1400
Precision: 1.0, recall: 0.05320361784601353
dataset: census, model: random forest (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3661.98s (training model: 16.66s, training explainer: 0.12s, experiment: 3645.19s)
Number of extracted samples: 1428
Number of accurate extracted samples: 1428
Precision: 1.0, recall: 0.0542676902029338
dataset: census, model: random forest (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3606.43s (training model: 16.51s, training explainer: 0.12s, experiment: 3589.80s)
Number of extracted samples: 1392
Number of accurate extracted samples: 1392
Precision: 1.0, recall: 0.05289959717260774
dataset: census, model: random forest (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3588.89s (training model: 16.45s, training explainer: 0.13s, experiment: 3572.31s)
Number of extracted samples: 1442
Number of accurate extracted samples: 1442
Precision: 1.0, recall: 0.05479972638139394
dataset: census, model: random forest (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 3582.40s (training model: 16.66s, training explainer: 0.12s, experiment: 3565.61s)
Number of extracted samples: 1421
Number of accurate extracted samples: 1421
Precision: 1.0, recall: 0.054001672113703734
dataset: census, model: neural network (repetition 0)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1932.58s (training model: 54.60s, training explainer: 0.13s, experiment: 1877.86s)
Number of extracted samples: 1885
Number of accurate extracted samples: 1885
Precision: 1.0, recall: 0.07163487117123965
dataset: census, model: neural network (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 2063.71s (training model: 53.73s, training explainer: 0.13s, experiment: 2009.85s)
Number of extracted samples: 1876
Number of accurate extracted samples: 1876
Precision: 1.0, recall: 0.07129284791365813
dataset: census, model: neural network (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1931.95s (training model: 54.31s, training explainer: 0.12s, experiment: 1877.52s)
Number of extracted samples: 1765
Number of accurate extracted samples: 1765
Precision: 1.0, recall: 0.06707456107015278
dataset: census, model: neural network (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1912.46s (training model: 54.09s, training explainer: 0.13s, experiment: 1858.24s)
Number of extracted samples: 1928
Number of accurate extracted samples: 1928
Precision: 1.0, recall: 0.07326898229079577
dataset: census, model: neural network (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1957.18s (training model: 64.30s, training explainer: 0.17s, experiment: 1892.71s)
Number of extracted samples: 1857
Number of accurate extracted samples: 1857
Precision: 1.0, recall: 0.07057079881431938
dataset: census, model: neural network (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1862.12s (training model: 54.45s, training explainer: 0.12s, experiment: 1807.54s)
Number of extracted samples: 2010
Number of accurate extracted samples: 2010
Precision: 1.0, recall: 0.07638519419320514
dataset: census, model: neural network (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1905.53s (training model: 54.40s, training explainer: 0.13s, experiment: 1851.01s)
Number of extracted samples: 1900
Number of accurate extracted samples: 1900
Precision: 1.0, recall: 0.0722049099338755
dataset: census, model: neural network (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1936.91s (training model: 56.75s, training explainer: 0.12s, experiment: 1880.03s)
Number of extracted samples: 1892
Number of accurate extracted samples: 1892
Precision: 1.0, recall: 0.07190088926046971
dataset: census, model: neural network (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 1951.41s (training model: 54.13s, training explainer: 0.11s, experiment: 1897.18s)
Number of extracted samples: 1964
Number of accurate extracted samples: 1964
Precision: 1.0, recall: 0.07463707532112183
dataset: census, model: neural network (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 2007.82s (training model: 54.65s, training explainer: 0.13s, experiment: 1953.05s)
Number of extracted samples: 1865
Number of accurate extracted samples: 1865
Precision: 1.0, recall: 0.07087481948772517
dataset: census numeric, model: decision tree (repetition 0)
Total time: 115.57s (training model: 0.03s, training explainer: 0.00s, experiment: 115.54s)
Number of extracted samples: 1340
Number of accurate extracted samples: 1340
Precision: 1.0, recall: 0.1300592060564884
dataset: census numeric, model: decision tree (repetition 1)
Total time: 115.22s (training model: 0.02s, training explainer: 0.00s, experiment: 115.20s)
Number of extracted samples: 1380
Number of accurate extracted samples: 1380
Precision: 1.0, recall: 0.13394157041638358
dataset: census numeric, model: decision tree (repetition 2)
Total time: 115.26s (training model: 0.02s, training explainer: 0.02s, experiment: 115.23s)
Number of extracted samples: 1427
Number of accurate extracted samples: 1427
Precisi

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 175.43s (training model: 11.43s, training explainer: 0.02s, experiment: 163.98s)
Number of extracted samples: 1216
Number of accurate extracted samples: 1216
Precision: 1.0, recall: 0.11802387654081335
dataset: census numeric, model: neural network (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 184.49s (training model: 11.64s, training explainer: 0.00s, experiment: 172.84s)
Number of extracted samples: 1480
Number of accurate extracted samples: 1480
Precision: 1.0, recall: 0.1436474813161215
dataset: census numeric, model: neural network (repetition 2)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 179.59s (training model: 11.45s, training explainer: 0.02s, experiment: 168.12s)
Number of extracted samples: 1203
Number of accurate extracted samples: 1203
Precision: 1.0, recall: 0.11676210812384742
dataset: census numeric, model: neural network (repetition 3)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 178.62s (training model: 11.50s, training explainer: 0.00s, experiment: 167.11s)
Number of extracted samples: 972
Number of accurate extracted samples: 972
Precision: 1.0, recall: 0.09434145394545278
dataset: census numeric, model: neural network (repetition 4)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 175.44s (training model: 11.59s, training explainer: 0.00s, experiment: 163.85s)
Number of extracted samples: 1224
Number of accurate extracted samples: 1224
Precision: 1.0, recall: 0.1188003494127924
dataset: census numeric, model: neural network (repetition 5)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 185.85s (training model: 11.42s, training explainer: 0.00s, experiment: 174.43s)
Number of extracted samples: 1199
Number of accurate extracted samples: 1199
Precision: 1.0, recall: 0.1163738716878579
dataset: census numeric, model: neural network (repetition 6)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 180.17s (training model: 12.28s, training explainer: 0.02s, experiment: 167.87s)
Number of extracted samples: 965
Number of accurate extracted samples: 965
Precision: 1.0, recall: 0.09366204018247112
dataset: census numeric, model: neural network (repetition 7)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 178.26s (training model: 12.06s, training explainer: 0.00s, experiment: 166.20s)
Number of extracted samples: 1393
Number of accurate extracted samples: 1393
Precision: 1.0, recall: 0.1352033388333495
dataset: census numeric, model: neural network (repetition 8)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 180.30s (training model: 12.27s, training explainer: 0.00s, experiment: 168.04s)
Number of extracted samples: 1207
Number of accurate extracted samples: 1207
Precision: 1.0, recall: 0.11715034455983694
dataset: census numeric, model: neural network (repetition 9)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Total time: 180.24s (training model: 12.37s, training explainer: 0.00s, experiment: 167.87s)
Number of extracted samples: 1161
Number of accurate extracted samples: 1161
Precision: 1.0, recall: 0.11268562554595749
dataset: census categorical, model: decision tree (repetition 0)
Total time: 1043.56s (training model: 0.53s, training explainer: 0.19s, experiment: 1042.84s)
Number of extracted samples: 14220
Number of accurate extracted samples: 14220
Precision: 1.0, recall: 0.788860534783091
dataset: census categorical, model: decision tree (repetition 1)
Total time: 1047.61s (training model: 0.53s, training explainer: 0.17s, experiment: 1046.91s)
Number of extracted samples: 14153
Number of accurate extracted samples: 14153
Precision: 1.0, recall: 0.7851436813491623
dataset: census categorical, model: decision tree (repetition 2)
Total time: 1039.98s (training model: 0.56s, training explainer: 0.20s, experiment: 1039.22s)
Number of extracted samples: 14057
Number of accurate extracted sa

c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


Total time: 1896.08s (training model: 11.03s, training explainer: 0.17s, experiment: 1884.88s)
Number of extracted samples: 13927
Number of accurate extracted samples: 13927
Precision: 1.0, recall: 0.7726062354377011
dataset: census categorical, model: random forest (repetition 1)


c:\users\ml_ks\venvs\xai_new\lib\site-packages\sklearn\ensemble\_forest.py:427: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


# Results

Precision is the percentage of extracted samples that is actually from the training data. 

Recall is the ratio of the number extracted training samples to all training samples.

In [ ]:
results

In [ ]:
file_name = 'results/2-1-cf-training-data-extraction-results'
if DATASET_HALF:
    file_name += '_dataset_size_halved'
results.to_csv(file_name + '.csv', index=False, na_rep='NaN', float_format='%.3f')

# Discussion

In our experiments, training data extraction with counterfactuals drawn from the training data has a recall between 45% and 67% for numeric data and 30% to 64% for categorical data. Since the attack cannot produce any false positive samples, precision is always 100%.